punchcut/src/source/glyf/scaler.rs

```diff
@@ -0,0 +1,640 @@
+use super::math;
+use super::{Context, Outline, Point};
+use crate::{Error, Hinting, NormalizedCoord, Result};
+
+use read_fonts::tables::{
+    glyf::{Anchor, CompositeGlyph, CompositeGlyphFlags, Glyf, Glyph, SimpleGlyph},
+    hmtx::Hmtx,
+    hvar::Hvar,
+    loca::Loca,
+};
+use read_fonts::types::{BigEndian, F2Dot14, GlyphId, Tag};
+use read_fonts::TableProvider;
+
+/// Limit for recursion when loading composite glyphs.
+const RECURSION_LIMIT: usize = 32;
+
+/// TrueType glyph scaler for a specific font and configuration.
+pub struct Scaler<'a> {
+    /// Backing context.
+    context: &'a mut Context,
+    /// Current font data.
+    font: Font<'a>,
+    /// Font identifier for the hinting cache.
+    font_id: Option<u64>,
+    /// Current hinting cache slot.
+    cache_slot: Option<CacheSlot>,
+    /// True if the outline is begin scaled.
+    is_scaled: bool,
+    /// Size in pixels per em unit.
+    ppem: u16,
+    /// Scale factor from font units to 26.6 fixed point.
+    scale: i32,
+    /// Hint mode.
+    hint: Option<Hinting>,
+    /// Normalized variation coordinates.
+    coords: &'a [NormalizedCoord],
+}
+
+impl<'a> Scaler<'a> {
+    /// Creates a new scaler for extracting outlines with the specified font
+    /// and configuration.
+    pub fn new(
+        context: &'a mut Context,
+        font: &impl TableProvider<'a>,
+        font_id: Option<u64>,
+        size: f32,
+        hint: Option<Hinting>,
+        coords: &'a [NormalizedCoord],
+    ) -> Result<Self> {
+        let font = Font::new(font)?;
+        let size = size.abs();
+        let ppem = size as u16;
+        let upem = font.units_per_em;
+        let (is_scaled, scale) = if size != 0. && upem != 0 {
+            (true, math::div((size * 64.) as i32, upem as i32))
+        } else {
+            (false, 0)
+        };
+        Ok(Self {
+            context,
+            font,
+            font_id,
+            cache_slot: None,
+            is_scaled,
+            ppem,
+            scale,
+            hint,
+            coords,
+        })
+    }
+
+    /// Loads an outline for the specified glyph identifier to the preallocated
+    /// target.
+    pub fn get_into(&mut self, glyph_id: GlyphId, outline: &mut Outline) -> Result<()> {
+        outline.clear();
+        self.context.unscaled.clear();
+        self.context.original.clear();
+        self.context.deltas.clear();
+        if glyph_id.to_u16() >= self.font.glyph_count {
+            return Err(Error::GlyphNotFound(glyph_id));
+        }
+        outline.is_scaled = self.is_scaled;
+        GlyphScaler::new(self).load(glyph_id, outline, 0)
+    }
+
+    /// Loads an outline for the specified glyph identifier.
+    pub fn get(&mut self, glyph_id: GlyphId) -> Result<Outline> {
+        let mut outline = Outline::new();
+        self.get_into(glyph_id, &mut outline)?;
+        Ok(outline)
+    }
+}
+
+/// State for loading a glyph.
+struct GlyphScaler<'a, 'b> {
```

Comment:
just 'a I can handle but if you are going to have more than 1 I feel strongly names and/or commentary makes sense.

Reply:
Will make this more clear.